coins/monero/src/wallet/send/mod.rs

```diff
@@ -197,6 +184,55 @@ async fn prepare_inputs<R: RngCore + CryptoRng, RPC: RpcConnection>(
   Ok(signable)
 }
 
+/// Deterministically calculate what the tx weight and fee will be
+fn calculate_weight_and_fee(
+  protocol: Protocol,
+  decoy_weights: &Vec<usize>,
+  n_inputs: usize,
+  n_outputs: usize,
+  extra: usize,
+  fee_rate: Fee,
+) -> (usize, u64) {
+  let mut weight: usize = 0;
+  let mut fee: u64 = 0;
+  // Starting the fee at 0 here is different than core Monero's wallet2.cpp,
+  // which starts its fee calculation with an estimate. This difference is ok in
+  // practice because wallet2 still ends up using a fee calculated from a tx's weight,
+  // as we calculate later in this function. See this PR highlighting wallet2's behavior:
+  // https://github.com/monero-project/monero/pull/8882
+  // Even with that PR, if the estimated fee's varint byte length is larger than
+  // the calculated fee's, the wallet can theoretically use a fee not
+  // based on the *final* tx weight. However, this does not occur in practice:
+  // it's nearly impossible for wallet2 to estimate a fee that is larger
+  // than the calculated fee today, and on top of that, even more unlikely for
+  // that estimate varint to be larger in byte length than the calculated fee.
+
+  let mut done = false;
+  let mut iters = 0;
+  let max_iters = 5;
+  while !done {
+    weight = Transaction::fee_weight(protocol, decoy_weights, n_inputs, n_outputs, extra, fee);
+
+    let fee_calculated_from_weight = fee_rate.calculate_fee_from_weight(weight);
+
+    // Continue trying to use the fee calculated from the tx's weight
+    done = fee_calculated_from_weight == fee;
+
+    fee = fee_calculated_from_weight;
+
+    debug_assert!(iters < max_iters, "Reached max fee calculation attempts");
```

Comment:
This isn't an actual invariant and shouldn't be a debug assert.

Reply:
This is there to pick up an unexpected result. It should never happen unless there is a mistake in the code. Plus I figure an actual invariant should be an assert. What is `debug_assert` for if not for a situation like this?

Reply:
Agreed it should never happen. Statement as an invariant requires effective certainty this won't be hit, no matter the arguments/variables at play. We allow custom fee rate specification, and that introduces a massive surface scope.

Unlikely != impossible without error.

You can likely turn this into an invariant by stating that the fee will only go up, and will stop going up once the VarInt hits its maximum length in bytes. Additionally, any increases in the fee requires the VarInt having gone up in bytes, and accordingly, the maximum execution depth is 10 (preventing endless execution). But then the max_iters has to be the execution depth which it should already halt on, making this debug_assert pointless.

The reasons to remove this debug_assert, and unbound iterations which may allow ridiculous values, are:

1) So we can keep other debug_asserts
2) Because the fee already has to be inspected before signing continues

If the concern is programs not doing #2, we should emphasize the API around doing so like Timelock.

As for this specific debug_assert, you can rewrite as
```
#[cfg(test)]
debug_assert
```
So it's caught in testing yet will never effect actual programs, compiled with debug_asserts or not.